### PR TITLE
Call subscription.complete on interpreter stop

### PR DESCRIPTION
**Pull request by pixtron**
When an interpreter is stoped, [all listeners get removed](https://github.com/statelyai/xstate/blob/6badd2ba3642391bee640aa4914003ad57f2e703/packages/core/src/interpreter.ts#L559-L562), therefore a potential subscription should complete.

Closes: #3452

### CODE REVIEW

File: packages/core/src/interpreter.ts [hunk 1]
```diff
@@ -112,6 +112,29 @@ export enum InterpreterStatus {
...
+    }
+    this.fired = true;
+    if (arguments.length === 0) {
+      return this.listener.call(this.target);
+    }
+    return this.listener.apply(this.target, arguments);
+  }
+}
+
+function _onceWrap(target: any, listener: () => void) {
```
Comment (pixtron) [inline packages/core/src/interpreter.ts:130]: not sure how to type `target` which should be an instance of `Interpreter`

File: packages/core/src/interpreter.ts [hunk 2]
```diff
@@ -388,8 +388,8 @@ export class Interpreter<
     return this;
   }
   public subscribe(
-    observer: Observer<
-      State<TContext, TEvent, any, TTypestate, TResolvedTypesMeta>
+    observer: Partial<
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:391]: I've rechecked with RxJS and listener types are actually all optional (well, not exactly, but I'm not sure if we want to do [this](https://github.com/ReactiveX/rxjs/blob/9a85e01625a1ee7a23e4ab2c2c6d3089a58e2e54/src/internal/types.ts#L170)).

It was really annoying in tests for me to implement this:
```js
service.subscribe({
  next: noop,
  error: noop,
  complete: completeCb
})
```
and I think that it makes sense to make it easier to just provide a single callback type, so I've followed that instinct and wrapped `Observer` at the parameter position `Partial`

File: packages/core/src/interpreter.ts [hunk 3]
```diff
@@ -404,48 +404,39 @@ export class Interpreter<
...
+      this.doneListeners.delete(completeOnce);
+      this.stopListeners.delete(completeOnce);
+      observer.complete();
+    };
+
+    if (this.status === InterpreterStatus.Stopped) {
+      observer.complete();
+    } else {
+      this.onDone(completeOnce);
+      this.onStop(completeOnce);
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:432]: this was a little bit annoying, but I think it's the cleanest solution possible right now. This probably will be completely refactored in v5 anyway.

File: packages/core/src/utils.ts [hunk 4]
```diff
@@ -699,20 +699,20 @@ export function toInvokeSource(
...
 
   return {
-    next: nextHandler,
-    error: errorHandler || noop,
-    complete: completionHandler || noop
+    next: ((isObserver ? nextHandler.next : nextHandler) || noop).bind(self),
+    error: ((isObserver ? nextHandler.error : errorHandler) || noop).bind(self),
+    complete: (
+      (isObserver ? nextHandler.complete : completionHandler) || noop
+    ).bind(self)
```
Comment (Andarist) [inline packages/core/src/utils.ts:715]: I don't think that we really need those `bind`s here but we were already binding `next` listener in the `Interpreter#subscribe`. So I didn't want to regress this accidentally and I've just used the same code pattern for all the callbacks here.

**Review by mellson — APPROVED** [review]
The code looks top quality 👍🏻

**Pull request merged**